### PR TITLE
Update code and move modeling.py to utils folder.

### PR DESCRIPTION
**Pull request by PenghuiCheng**
# What does this PR do?

Move modeling.py to optimum.intel.utils folder and update the tracing code.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/intel/utils/modeling.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+                "use_auth_token": use_auth_token,
+                "cache_dir": cache_dir,
+                "subfolder": subfolder,
+                "local_files_only": local_files_only,
+                "force_download": force_download,
+            }
+            model = TasksManager.get_model_from_task(task, model_id, **model_kwargs)
+            model_inputs = None
+        else:
+            return cls._from_transformers(
```
Comment (echarlaix) [inline optimum/intel/utils/modeling.py:236]: I would rather keep the `export` parameter instead of `torchscript` to keep consistency accross all the `OptimizedModel`. To make it more explicit, why not change the class name instead ? To something like `TorchScriptModelForCausalLM` ?

File: optimum/intel/utils/modeling.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+
+            self.generation_config = GenerationConfig.from_model_config(config)
+
+    @staticmethod
+    def load_model(file_name: Union[str, Path]):
+        model = torch.jit.load(file_name)
+        torch.jit.freeze(model.eval())
+        return model
+
+    def set_traced_model(self, traced_model):
```
[the remark this answers is not in the document]
Comment (PenghuiCheng) [reply optimum/intel/utils/modeling.py:134]: This function is for quantization with INC, for example:
model = AutoModelForGeneration.from_pretrained(args.model_name_or_path, torchscript=True)

def eval_function(q_model):
    model.set_traced(model)
    for input in dataloader:
        out=model(**input)
    eval_metric = metric(out, label)
    return eval_metric

q_model = inc_quantizer.quantize(model, config, calib_dataloader, eval_func=eval_function)

File: optimum/intel/utils/modeling.py (hunk 1 above)
Comment (PenghuiCheng) [reply optimum/intel/utils/modeling.py:236]: I have retained the export parameter, but I don't think the semantics of export match the tracing jit model very well, so I added the torchscript parameter. It's ok for me to change the class name, TorchScriptModelForCausalLM is OK.

Comment (PenghuiCheng) [reply optimum/intel/utils/modeling.py:236]: How about TorchScriptModelForGeneration? because not all generation model is CausalLM, like gpt2 is GPT2LMHeadModel

File: optimum/intel/generation/modeling.py [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+            model_kwargs = {
+                "revision": revision,
+                "use_auth_token": use_auth_token,
+                "cache_dir": cache_dir,
+                "subfolder": "",
+                "local_files_only": local_files_only,
+                "force_download": force_download,
+            }
+
+            model = TasksManager.get_model_from_task(task, "../../ipex/text-generation/model", **model_kwargs)
```
Comment (echarlaix) [inline optimum/intel/generation/modeling.py:204]: Shouldn't this be removed 
```suggestion

```

File: optimum/intel/generation/modeling.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+                model_id=model_id,
+                config=config,
+                use_auth_token=use_auth_token,
+                revision=revision,
+                force_download=force_download,
+                cache_dir=cache_dir,
+                local_files_only=local_files_only,
+                use_cache=use_cache,
+                **kwargs,
+            )
```
Comment (echarlaix) [inline optimum/intel/generation/modeling.py:254]: When would this case happen ? `_from_transformers` should be called when `export=True`

```suggestion
```

File: optimum/intel/generation/modeling.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+            model_kwargs = {
+                "revision": revision,
+                "use_auth_token": use_auth_token,
+                "cache_dir": cache_dir,
+                "subfolder": subfolder,
+                "local_files_only": local_files_only,
+                "force_download": force_download,
+            }
+            model = TasksManager.get_model_from_task(task, model_id, **model_kwargs)
+            model_inputs = None
```
Comment (echarlaix) [inline optimum/intel/generation/modeling.py:242]: In which case would we need the original model ?

```suggestion
```

File: optimum/intel/generation/modeling.py [hunk 6]
```diff
@@ -204,16 +302,13 @@ def _from_transformers(
         save_dir = TemporaryDirectory()
         save_dir_path = Path(save_dir.name)
         torch.jit.save(traced_model, save_dir_path / WEIGHTS_NAME)
+        config.torchscript = True
 
-        return cls._from_pretrained(
```
Comment (echarlaix) [inline optimum/intel/generation/modeling.py:208]: I think we should keep `cls._from_pretrained` in order to have all the model transformation at one place (like  the 2 iterations to convert model to int8 model for IPEX)

File: optimum/intel/generation/modeling.py [hunk 7]
```diff
@@ -186,8 +277,15 @@ def _from_transformers(
             "force_download": force_download,
         }
 
-        model = TasksManager.get_model_from_task(task, model_id, **model_kwargs)
-        model.config.return_dict = False
+        if isinstance(model_id, str):
+            model = TasksManager.get_model_from_task(task, model_id, **model_kwargs)
+            model.config.return_dict = False
+        elif isinstance(model_id, torch.nn.Module):
```
Comment (echarlaix) [inline optimum/intel/generation/modeling.py:283]: I don't think we should leave the possibility to give all the options to give as input the model path, an `nn.Module` and a `RecursiveScriptModule` but to only give the model path. This could be confusing to users in my opinion, also this doesn't respect the format we have for Optimum `OVModel` and `ORTModel`

File: optimum/intel/generation/modeling.py [hunk 8]
```diff
@@ -108,6 +105,94 @@ def forward(
 
         return CausalLMOutputWithPast(logits=outputs[0], past_key_values=outputs[1] if self.use_cache else None)
 
+    @classmethod
+    @add_start_docstrings(FROM_PRETRAINED_START_DOCSTRING)
+    def from_pretrained(
```
[the remark this answers is not in the document]
Comment (PenghuiCheng) [reply optimum/intel/generation/modeling.py:110]: removed

**Review by echarlaix — APPROVED** [review]
LGTM, thanks a lot for iterating on this @PenghuiCheng  !!
